src/libutil/url.cc

```diff
@@ -408,21 +408,75 @@ ParsedUrlScheme parseUrlScheme(std::string_view scheme)
     };
 }
 
-ParsedURL fixGitURL(const std::string & url)
+struct ScpLike
 {
-    std::regex scpRegex("([^/]*)@(.*):(.*)");
-    if (!hasPrefix(url, "/") && std::regex_match(url, scpRegex))
-        return parseURL(std::regex_replace(url, scpRegex, "ssh://$1@$2/$3"));
+    ParsedURL::Authority authority;
+    std::string_view path;
+};
+
+/**
+ * Parse a scp url. This is a helper struct for fixGitURL.
+ * This is needed since we support scp-style urls for git urls.
+ * https://git-scm.com/book/ms/v2/Git-on-the-Server-The-Protocols
+ *
+ * A good reference is libgit2 also allows scp style
+ * https://github.com/libgit2/libgit2/blob/58d9363f02f1fa39e46d49b604f27008e75b72f2/src/util/net.c#L806
+ */
+static std::optional<ScpLike> parseScp(std::string_view s) noexcept
+{
+    if (s.empty() || s.front() == '/')
+        return std::nullopt;
+
+    // Find the colon that separates host from path.
+    // Find the right-most since ipv6 has colons
+    const auto colon = s.rfind(':');
+    if (colon == std::string_view::npos)
+        return std::nullopt;
+
+    // Split head:[path]
+    const auto head = s.substr(0, colon);
+    const auto path = s.substr(colon + 1);
+
+    if (head.empty())
+        return std::nullopt;
+
+    return ScpLike{
+        .authority = ParsedURL::Authority::parse(head),
+        .path = std::move(path),
+    };
+}
+
+ParsedURL fixGitURL(std::string_view url)
+{
+    if (auto r = boost::urls::parse_uri(url); r && r->has_scheme() && r->has_authority()) {
+        return parseURL(url);
+    }
+
     if (hasPrefix(url, "file:"))
         return parseURL(url);
-    if (url.find("://") == std::string::npos) {
```

Comment:
this can all be

```c++
try {
    return parseURL(url);
} catch (BadIURL &) {};
```
and then do the fallback logic below

Reply:
These guards are important. Without it parseURL successfully parses some SCP styles although its weird; the ports are the paths or something.

I think it's a bug we can open.

Reply:
That is just approximate off the top of my head, but the basic idea is we should not need to parse twice.